src/capymoa/classifier/_shrubs_ensemble.py

```diff
@@ -0,0 +1,306 @@
+            step_size = 1.0 / (len(self.estimators_) + 1.0)
+        else:
+            step_size = self.step_size
+
+        # Get all predictions of all trees for gradient computation
+        all_proba = self._individual_proba(data)
+
+        for i in range(self.burnin_steps + 1):
+            output = np.array([w * p for w,p in zip(all_proba, self.estimator_weights_)]).sum(axis=0)
+            # Compute derivative of the loss
+            if self.loss == "mse":
+                if self.n_classes_ == 1:
+                    loss_deriv = 2 * (output - target)
+                else:
+                    # Use one hot encoding in classification setting
+                    target_one_hot = np.array( [ [1.0 if y == i else 0.0 for i in range(self.n_classes_)] for y in target] )
+                    loss_deriv = 2 * (output - target_one_hot)
+                # loss = (output - target_one_hot) * (output - target_one_hot)
+            elif self.loss == "ce":
+                target_one_hot = np.array( [ [1.0 if y == i else 0.0 for i in range(self.n_classes_)] for y in target] )
+                m = target.shape[0]
+                loss_deriv = softmax(output, axis=1)
+                loss_deriv[range(m),target_one_hot.argmax(axis=1)] -= 1
+                # loss = -target_one_hot*np.log(softmax(output, axis=1) + 1e-7)
+            elif self.loss == "h2":
+                target_one_hot = np.array( [ [1.0 if y == i else -1.0 for i in range(self.n_classes_)] for y in target] )
+                zeros = np.zeros_like(target_one_hot)
+                loss_deriv = - 2 * target_one_hot * np.maximum(1.0 - target_one_hot * output, zeros) 
+                # loss = np.maximum(1.0 - target_one_hot * output, zeros)**2
+            else:
+                raise RuntimeError(f"The loss {self.loss} is unknown and for some reason you ended up in a code-path that should not be reachable. Currently only the losses {{ce, mse, h2}} are supported for classification. For regression only mse is supported. ")
+        
+            # Compute the gradient for the loss
+            directions = np.mean(all_proba*loss_deriv,axis=(1,2))
+
+            # Compute the gradient for the tree regularizer
+            if self.l_tree_reg != 0:
+                node_deriv = self.l_tree_reg * np.array([ est.tree_.node_count for est in self.estimators_ if isinstance(est, DecisionTreeClassifier)])
+            else:
+                node_deriv = 0
+
+            # Perform the gradient step. Note that L0 / L1 regularizer is performed via the prox operator and thus performed _after_ this update.
+            self.estimator_weights_ = self.estimator_weights_ - step_size*directions - step_size*node_deriv - step_size*self.l_l2_reg*2*self.estimator_weights_
+
+            # The latest tree should only receive updates in the last round of burn-in. Reset its weight here. This is somewhat arbitrary, but lead to better performance in our initial experiments. 
+            # if i < self.burnin_steps:
```

Comment:
Is the burn-in still being treated correctly

Reply:
This is a bit more tricky, as both versions are okay. I re-checked my original c++ implementation (which we used for the experiments in the paper) and there we performed burn-in for all trees (i.e. the version that it is now) equally. Hence, this version is closer to the original experiments. I would keep the code but maybe adept the comment accordingly.

Reply:
I decided to remove the commented code (line:255) and associated comment to be more consistent with your C++ implementation. Let me know if I got this the wrong way around and I'll patch it.